core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java

```diff
@@ -310,7 +403,80 @@ public void testThatForceRefreshBalksWhenAForceRefreshIsInProgress() throws Exce
   }
 
   @Test
-  public void testGetPreferredIpTypes() throws Exception {
+  public void testRefreshRetriesOnAfterFailedAttempts() throws Exception {
+    InstanceData aboutToExpireData = newInstanceDataExpiringIn(10, ChronoUnit.MILLIS);
+
+    InstanceData data = newFutureInstanceData();
+
+    AtomicInteger refreshCount = new AtomicInteger();
+    final PauseCondition badRequest1 = new PauseCondition();
+    final PauseCondition badRequest2 = new PauseCondition();
```

Comment:
Why two bad requests? Could we write this test with just one?

Reply:
Because I want to ensure that the retry is working - that the chain of futures is being built correctly even after the first failure. Early iterations of this PR had a bug where it would retry after one failure, but stop after the second failure.